pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/PartitionAwareOfflineRoutingTableBuilder.java

```diff
@@ -0,0 +1,186 @@
+import com.linkedin.pinot.common.metadata.segment.SegmentPartitionMetadata;
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadata;
+import com.linkedin.pinot.common.response.ServerInstance;
+import com.linkedin.pinot.transport.common.SegmentId;
+import com.linkedin.pinot.transport.common.SegmentIdSet;
+
+public class PartitionAwareOfflineRoutingTableBuilder extends AbstractRoutingTableBuilder {
+  private static final String PARTITION_METADATA_PRUNER = "PartitionZKMetadataPruner";
+
+  Map<SegmentId, Map<Integer, ServerInstance>> _segmentId2ServersMapping = new HashMap<>();
+  AtomicReference<Map<SegmentId, Map<Integer, ServerInstance>>> _mappingRef = new AtomicReference<>(_segmentId2ServersMapping);
+  Map<SegmentId, SegmentZKMetadata> _segment2SegmentZkMetadataMap = new ConcurrentHashMap<>();
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private SegmentZKMetadataPrunerService _pruner;
+  private PartitionToReplicaGroupMappingZKMetadata _partitionToReplicaGroupMappingZKMedata;
+  private Random _random = new Random();
+  private TableConfig _tableConfig;
+  private int _numReplicas;
+  private boolean _isPartitionLevelReplicaGroupAssignment;
+
+  @Override
+  public void init(Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableConfig = tableConfig;
+    _propertyStore = propertyStore;
+    _pruner = new SegmentZKMetadataPrunerService(new String[]{PARTITION_METADATA_PRUNER});
+
+    final String partitionColumn = _tableConfig.getValidationConfig().getReplicaGroupStrategyConfig().getPartitionColumn();
+    _isPartitionLevelReplicaGroupAssignment = (partitionColumn == null) ? true : false;
+  }
+
+  @Override
+  public void computeRoutingTableFromExternalView(String tableName, ExternalView externalView, List<InstanceConfig> instanceConfigList) {
+    RoutingTableInstancePruner pruner = new RoutingTableInstancePruner(instanceConfigList);
+    String[] segmentSet = externalView.getPartitionSet().toArray(new String[externalView.getPartitionSet().size()]);
+    _numReplicas = _tableConfig.getValidationConfig().getReplicationNumber();
+    _partitionToReplicaGroupMappingZKMedata = ZKMetadataProvider.getPartitionToReplicaGroupMappingZKMedata(_propertyStore, tableName);
+
+    // Compute partition id set
+    Set<Integer> partitionIds = new HashSet<>();
+    for (String segment : segmentSet) {
+      SegmentId segmentId = new SegmentId(segment);
+      // retrieve the metadata for the segment and compute the partitionIds set
+      SegmentZKMetadata segmentZKMetadata = _segment2SegmentZkMetadataMap.get(segmentId);
+      if (segmentZKMetadata == null) {
+        segmentZKMetadata = ZKMetadataProvider.getOfflineSegmentZKMetadata(_propertyStore, tableName, segment);
+        _segment2SegmentZkMetadataMap.put(segmentId, segmentZKMetadata);
+      }
+      int partitionId = getPartitionId(segmentZKMetadata);
```

Comment:
If there is no partition information, I sequentially assign replica id for each server and add it to look up table.